gem/impero_utils.py

```diff
@@ -61,17 +60,11 @@ def nonzero(assignment):
     expressions = [expression for variable, expression in assignments]
 
     # Collect indices in a deterministic order
-    indices = OrderedSet()
-    for node in traversal(expressions):
-        if isinstance(node, gem.Indexed):
-            for index in node.multiindex:
-                if isinstance(index, gem.Index):
-                    indices.add(index)
-        elif isinstance(node, gem.FlexiblyIndexed):
-            for offset, idxs in node.dim2idxs:
-                for index, stride in idxs:
-                    if isinstance(index, gem.Index):
-                        indices.add(index)
+    indices = list(collections.OrderedDict.fromkeys(chain.from_iterable(
+        node.index_ordering()
+        for node in traversal(expressions)
+        if isinstance(node, (gem.Indexed, gem.FlexiblyIndexed))
+    )))
```

Comment:
OK, simplifying existing code.  You collect unique indices in the order they appear in the dag traversal.